api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/AddPatientToQueue.java

```diff
@@ -0,0 +1,75 @@
+import org.openmrs.module.queue.api.QueueServicesWrapper;
+import org.openmrs.module.queue.api.search.QueueSearchCriteria;
+import org.openmrs.module.queue.model.Queue;
+import org.openmrs.module.queue.model.QueueEntry;
+
+public class AddPatientToQueue implements org.openmrs.module.htmlformentry.CustomFormSubmissionAction {
+
+    private static final String ADD_TO_QUEUE_CONCEPT_CODE_PIH = "1272";
+    protected Log log = LogFactory.getLog(getClass());
+
+    @Override
+    public void applyAction(FormEntrySession formEntrySession) {
+
+        Concept addToQueueConcept = Context.getConceptService().getConceptByMapping(ADD_TO_QUEUE_CONCEPT_CODE_PIH, "PIH");
+
+        if (addToQueueConcept == null) {
+            log.error("Concept PIH: " + ADD_TO_QUEUE_CONCEPT_CODE_PIH + " cannot be found");
+            return;
+        }
+        Patient patient = formEntrySession.getPatient();
+        Encounter encounter = formEntrySession.getEncounter();
+
+        for (Obs candidate : encounter.getObsAtTopLevel(false)) {
+            if (candidate.getConcept().equals(addToQueueConcept)) {
+                // The coded answer represents the Concept associated with the queue service
+                Concept serviceQueue = candidate.getValueCoded();
+                QueueSearchCriteria criteria = new QueueSearchCriteria();
+                criteria.setServices(Arrays.asList(serviceQueue));
+                QueueServicesWrapper queueServicesWrapper = Context.getRegisteredComponents(QueueServicesWrapper.class).get(0);
+                List<Queue> queues = queueServicesWrapper.getQueueService().getQueues(criteria);
+                Queue queue = null ;
+                QueueEntry queueEntry = null;
+                if ( !queues.isEmpty() ) {
+                    queue = queues.get(0);
+                }
+                if (queue != null ) {
+                    List<Concept> allowedStatuses = queueServicesWrapper.getAllowedStatuses(queue);
+                    if (allowedStatuses.isEmpty()) {
+                        log.error("Queue " + queue.getName() + " has not valid allowed statuses");
+                        return;
+                    }
+                    List<Concept> allowedPriorities = queueServicesWrapper.getAllowedPriorities(queue);
+                    if (allowedPriorities.isEmpty()) {
+                        log.error("Queue " + queue.getName() + " has not valid allowed priorities");
+                        return;
+                    }
+                    queueEntry = new QueueEntry();
+                    queueEntry.setQueue(queue);
+                    queueEntry.setPatient(patient);
+                    // The first entry in the allowedPriorities should always be Normal
+                    queueEntry.setPriority(allowedPriorities.get(0));
+                    // The first entry in the allowedStatuses should always be Waiting
+                    queueEntry.setStatus(allowedStatuses.get(0));
```

Comment:
I don't think we should make this assumption.  I think we should just set priority to Normal and status to Waiting explicitly if that is what we want to see happen.  Then, if there are situations where we want something different, we can update the logic here explicitly to handle this.  This is in the pihcore module, and our metadata definition is in the pihemr config, we know what is valid and not valid, and we should fail hard if the expected situation is not found, as this reflects something wrong happened in the code when starting up.

Reply:
@mseaton , it is easy enough to just hard code the Status and Priority concepts, but we just did some configuration work on the queue module and the config-pihsl that allows us to configure each queue individually with different statuses and priorities. And you said that the best way to determine the allowed statuses and priorities for a given queue is to use the gerAllowed methods. We should ignore this possible custom configuration now, and just hardcode here the Normal and Waiting concepts? Then, maybe I should remove the serviceConcept too, and just hardcode to the MCOE Triage concept?